python/lsst/source/injection/combine.py

```diff
@@ -0,0 +1,323 @@
+                catalog_dict[band] = []
+            # load the patch number to check for patch overlap duplicates later
+            catalog["patch"] = ref.dataId.patch.id
+            catalog_dict[band].append(catalog)
+            tracts.add(ref.dataId.tract.id)
+        # vstack all the catalogs for each band
+        for band, catalog_list in catalog_dict.items():
+            catalog_dict[band] = vstack(catalog_list)
+        if len(tracts) != 1:
+            raise RuntimeError(f"len({tracts=}) != 1")
+        return (catalog_dict, list(tracts)[0])
+
+    def _remove_duplicates(
+        self,
+        catalog: Table,
+        tractInfo,
+    ) -> Table:
+        """Remove tract and patch overlap duplicates."""
+        # TODO: add isPatchInner column
+        duplicates = []
+        for ind, row in enumerate(catalog):
+            tractInner = tractInfo.getInnerSkyPolygon()
+            # check if the source is within inner tract bounds
+            if tractInner.contains(np.deg2rad(row[self.config.col_ra]), np.deg2rad(row[self.config.col_dec])):
+                spherePoint = SpherePoint(
+                    row[self.config.col_ra] * degrees, row[self.config.col_dec] * degrees
+                )
+                # get the patch number from the source's ra,dec
+                patchInfo = tractInfo.findPatch(spherePoint)
+                # check against the patch column
+                if row["patch"] != patchInfo.sequential_index:
+                    duplicates.append(ind)
+            # if the source is not within inner tract bounds, remove it
+            else:
+                duplicates.append(ind)
+        catalog.remove_rows(duplicates)
+        return catalog
+
+    # TODO: implement support for combining multi-component objects on DM-42433
+    def _combine_multi_component_objects(
+        self,
+        catalog: Table,
+        tractInfo,
+        match_radius: float,
+    ) -> Table:
+        # TODO: when creating an injection catalog, make a groupId column to
+        # accurately identify multi-component objects. Use that here instead
+        # matching.
+        with Matcher(catalog[self.config.col_ra], catalog[self.config.col_dec]) as m:
+            idx, selfMatchIndices, _, dists = m.query_self(
+                match_radius,
+                min_match=2,
+                return_indices=True,
+            )
+        # get the indices of matched groups of sources
+        matched_inds_groups = [list(i) for i in set(map(tuple, idx)) if len(i) != 0]
+        # combine matched sources
+        for inds in matched_inds_groups:
+            # create a new empty row
+            catalog.add_row()
+            # take the position of the brightest component
+            bright_index = inds[
+                np.where(catalog[inds][self.config.col_mag] == catalog[inds][self.config.col_mag].min())[0][0]
+            ]
+            catalog[-1][self.config.col_ra] = np.min(catalog[bright_index][self.config.col_ra])
```

Comment:
Why catalog[-1]?

Reply:
It just puts it at the end of the catalog. It was an arbitrary choice.